core/src/main/java/org/elasticsearch/tasks/CancellableTask.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.common.collect.Tuple;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A task that can be canceled
+ */
+public class CancellableTask extends Task {
+
+    private final AtomicReference<String> reason = new AtomicReference<>();
+
+    public CancellableTask(long id, String type, String action, String description) {
+        super(id, type, action, description);
+    }
+
+    public CancellableTask(long id, String type, String action, String description, String parentNode, long parentId) {
+        super(id, type, action, description, parentNode, parentId);
+    }
+
+    /**
+     * This method is called by the task manager when this task is cancelled.
+     */
+    final void cancel(String reason) {
+        assert reason != null;
+        if (this.reason.compareAndSet(null, reason)) {
+            doCancel(reason);
+        }
+    }
+
+    /**
+     * An action can override this method to be notified when the task is cancelled
+     */
+    protected void doCancel(String reason) {
```

Comment:
I am envisioning task to be used in two ways. An action can ask the task periodically, "am I cancelled yet?", or it can subscribe itself for cancellation notification and be more proactive in the way it handles it. The doCancel method support the second way of doing thing. This pattern is implemented in CancellableTaskWithNotification in one of the tests.